helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -126,4 +123,53 @@ public OptimalAssignment calculate(ClusterModel clusterModel) throws HelixRebala
     return hardConstraints.stream().map(HardConstraint::getDescription)
         .collect(Collectors.toList());
   }
+
+  // TODO investigate better ways to sort replicas. One option is sorting based on the creation time.
+  private List<AssignableReplica> getOrderedAssignableReplica(ClusterModel clusterModel) {
+    Map<String, Set<AssignableReplica>> replicasByResource = clusterModel.getAssignableReplicaMap();
+    List<AssignableReplica> orderedAssignableReplicas =
+        replicasByResource.values().stream().flatMap(replicas -> replicas.stream())
+            .collect(Collectors.toList());
+
+    Map<String, ResourceAssignment> bestPossibleAssignment =
+        clusterModel.getBestPossibleAssignment();
+    Map<String, ResourceAssignment> baselineAssignment = clusterModel.getBaseline();
+
+    // 1. Sort according if the assignment exists in the best possible and/or baseline assignment
+    // 2. Sort according to the state priority. Note that prioritizing the top state is required.
+    // Or the greedy algorithm will unnecessarily shuffle the states between replicas.
+    // 3. Sort according to the resource/partition name.
+    orderedAssignableReplicas.sort((o1, o2) -> {
```

Comment:
nit: the comparator can be definitely simplified, at least rename these to (replica1, replica2)

Reply:
It makes more sense now if we sort based on baseline & bestpossible assignment. The previous way of sorting it by alphabetical order doesn't make sense to me